### ISSUE TITLE
NullPointerException in validateObjectIfNotBorrowed(ObjectHolder.java:132)

### ISSUE THREAD

**Issue by kciryt** [zolyfarkas/spf4j#24]
I've been using this lib for a year or so, then yesterday I suddenly started getting an this null ptr exception.

```
x 11:35:58 x-3: 11:35:58.543 [vert.x-worker-thread-16] WARN  o.spf4j.recyclable.impl.ObjectHolder - Validation of com.x.MyClass@93e7651 failed, detail null 
x 11:35:58 x-3: 11:35:58.544 [vert.x-worker-thread-16] ERROR org.spf4j.base.AbstractRunnable - Exception in runnable: 
x 11:35:58 x-3: java.lang.NullPointerException: null
x 11:35:58 x-3: 	at org.spf4j.recyclable.impl.ObjectHolder.validateObjectIfNotBorrowed(ObjectHolder.java:132)
x 11:35:58 x-3: 	at org.spf4j.recyclable.impl.RecyclingSupplierBuilder$AbstractRunnableImpl$1.handle(RecyclingSupplierBuilder.java:116)
x 11:35:58 x-3: 	at org.spf4j.recyclable.impl.RecyclingSupplierBuilder$AbstractRunnableImpl$1.handle(RecyclingSupplierBuilder.java:113)
x 11:35:58 x-3: 	at org.spf4j.recyclable.impl.SimpleSmartObjectPool.scan(SimpleSmartObjectPool.java:304)
x 11:35:58 x-3: 	at org.spf4j.recyclable.impl.ScalableObjectPool.scan(ScalableObjectPool.java:70)
x 11:35:58 x-3: 	at org.spf4j.recyclable.impl.RecyclingSupplierBuilder$AbstractRunnableImpl.doRun(RecyclingSupplierBuilder.java:113)
x 11:35:58 x-3: 	at org.spf4j.base.AbstractRunnable.run(AbstractRunnable.java:80)
x 11:35:58 x-3: 	at com.x.MyClassPoolReturnScheduler.lambda$scheduleWithFixedDelay$0(MyClassPoolReturnScheduler.java:55)
x 11:35:58 x-3: 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:812)
x 11:35:58 x-3: 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:775)
x 11:35:58 x-3: 	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:337)
x 11:35:58 x-3: 	at io.vertx.core.impl.TaskQueue.lambda$new$0(TaskQueue.java:60)
x 11:35:58 x-3: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
x 11:35:58 x-3: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
x 11:35:58 x-3: at java.lang.Thread.run(Thread.java:745)
```

This seems very odd to me considering the code here ensured the object is never null:
 https://github.com/zolyfarkas/spf4j/blob/master/spf4j-core/src/main/java/org/spf4j/recyclable/impl/ObjectHolder.java#L128

Any ideas what could be causing this?  I am initializing my RecyclingSupplierBuilder with a factory that has the following dispose method:

```
@Override
public void dispose(MyClass myClass) throws ObjectDisposeException {
  if(myClass == null) {
    System.out.println("myClass is null?");
  }
  myClass.close();
}

```

However, I don't see the "myClass is null?" being printed--the null ptr exception seems to happen before.

**Comment by zolyfarkas** [comment]
what version of spf4j are you using?

**Comment by kciryt** [comment]
Using spf4j-core 8.2.11, Also I am noticing that it is not the obj that is null, but the factory.  When I debug I get:

```
object = {ObjectHolder@5954} "ObjectHolder{obj=com.x.MyClass@6d797e9, borrowed=false, factory=null}"
```

This is definitely confusing for me considering that factory.validate(obj, null) on line 122 was called fine above.

**Comment by zolyfarkas** [comment]
Interesting... factory should not be null.. it is also a final field... class is annotated with @ParametersAreNonNullByDefault... I will take a look and see how can this end up being null...

**Comment by zolyfarkas** [comment]
I looked at the latest code, (and unless I made a mistake) the only way that factory can be null, if the pool is constructed with is as null and passed null to  RecyclingSupplierBuilder...

**Comment by zolyfarkas** [comment]
I annotated the factory with @Nonull in the latest dev build...

**Comment by kciryt** [comment]
The RecyclingSupplierBuilder is definitely not being constructed with a null pool in my case.  I agree with you as far as the code looks.  I can't find any reason for the behavior I am seeing.  I will poke around some more and hopefully figure this out.

**Comment by zolyfarkas** [comment]
OK, found the issue, this happens when a object is borrowed from the pool in a different thread than it is returned from. (In conjunction with a maintenance thread). 
This should not be a common use case for this thread pool implementation because the implementation optimizes for the common case of borrowing and returning in the same thread.
Never the less it should work.

I pushed out a fix in maven central in with version 8.4.0-SNAPSHOT

8.4.0 should come out sometime next week...

**Comment by kciryt** [comment]
This is likely the issue.  We had just added a new library around our thread management--which seems to have triggered this exact scenario.  After we rolled back use of that lib, everything worked fine again.  Looking forward to 8.4.0.  Thanks!
